### PR TITLE
CI pipeline with Github Actions

### PR DESCRIPTION
**Pull request by diegoferigo**
I used this repository as a testbench for Github Actions. This is a first WIP multiplatform version I developed.

From these first tests, Github Actions are amazing :) The platform is very generic and it allows sharing actions through repositories. Given its generality, we should discuss f2f how to proceed organization-wise :tada:

cc @traversaro @pattacini @vtikha @claudiofantacci

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    - name: Environment Variables
+      shell: bash
+      run: env
+        
+    # ============
+    # DEPENDENCIES
+    # ============
+    
+    # Waiting a github action that installs dependencies in a
+    # multiplatform setup (likely using TS or JS), we install
```
Comment (traversaro) [inline .github/workflows/ci.yml]: What do `TS` and `JS` stand for?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      shell: bash
+      run: |
+        cd $GITHUB_WORKSPACE/../how-to-export-cpp-library
+        cd build
+        cmake --build . --config ${{ matrix.build_type }}
+
+    - name: Test
+      shell: bash
+      run: |
+        cd $GITHUB_WORKSPACE/../how-to-export-cpp-library
```
Comment (traversaro) [inline .github/workflows/ci.yml]: There is any way for making these `how-to-export-cpp-library` parametric in some way?

**Review by traversaro — DISMISSED** [review]

**Comment by pattacini** [comment]
Thanks @diegoferigo for these early tests 👍 

I had GitHub Actions on my radar since a while, knowing that it would have represented a real breakthrough once delivered.

However, my personal opinion is that iCub Tech should stay focused on a few "organizational" things in the upcoming weeks. After this start, we can come back to this.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (diegoferigo) [reply .github/workflows/ci.yml]: Typescript and Javascript. There are two types of Actions: docker-based and js-based. The former run isolated and can only operate on the `${GITHUB_WORKSPACE}` directory, the latter instead run in the host VM.

For instance, I tried to develop an Action to install dependencies in a multi-os setting (similar to the Travis addons). The only type to install deps that are persistent among steps is js, something I'm not really familiar with :/

File: .github/workflows/ci.yml (hunk 2 above)
Comment (diegoferigo) [reply .github/workflows/ci.yml]: Nah, it's a leftover. I used to clone this repo from another repo running this action, and I had to switch folder. It is not necessary in this case, the default initial directory of the steps is already that one.

**Comment by diegoferigo** [comment]
> However, my personal opinion is that iCub Tech should stay focus on a few "organizational" things in the upcoming weeks. After this start, we can come back to this.

I totally agree, this is not a priority at the moment. This PR was originated from just personal curiosity :)

**Comment by pattacini** [comment]
Go ahead @diegoferigo don't refrain your curiosity, absolutely!
I was only referring to our availability.

You may also propose a plan that could cover the organization and that we'll review later on.

**Review by claudiofantacci — DISMISSED** [review]

**Comment by claudiofantacci** [comment]
Very nice @diegoferigo. Not sure whether you are still trying something or not. If it is the case, could you please rename the PR with a prefix `[WIP]`? Otherwise, please squash the commits into one and let me know so that I can merge the PR right away! 🚀

**Comment by diegoferigo** [comment]
Let's wait a bit more before merging, I'd like to clean the file.

I'm not sure why the tests are not running in this PR. They work fine [on my fork](https://github.com/diegoferigo/how-to-export-cpp-library/runs/200487126).

**Comment by pattacini** [comment]
> Not sure whether you are still trying something or not. If it is the case, could you please rename the PR with a prefix `[WIP]`?

This may help out: https://github.blog/2019-02-14-introducing-draft-pull-requests.

**Comment by diegoferigo** [comment]
>This may help out: https://github.blog/2019-02-14-introducing-draft-pull-requests.

@pattacini I was not sure if CI is triggered on draft PRs.

**Comment by pattacini** [comment]
@diegoferigo some related topics:
- https://travis-ci.community/t/draft-pull-requests-not-being-built/2434/17
- https://changelog.travis-ci.com/draft-pull-requests-support-105121

According to the last one, Draft PR's should be built.

**Comment by diegoferigo** [comment]
I found [this documentation](https://help.github.com/en/articles/events-that-trigger-workflows#pull-request-events-for-forked-repositories) on pull request events for forked repositories. I suspect that if there is no Actions configuration upstream, the pipeline does not start.

**Comment by diegoferigo** [comment]
Cleaned and ready to merge. [This is](https://github.com/diegoferigo/how-to-export-cpp-library/runs/200900842) the pipeline running on my fork. Hopefully after merging it will be enabled also in this repository.

The badge link should be correct, though we'll discover it only after merging.

**Comment by traversaro** [comment]
> Cleaned and ready to merge. This is the pipeline running on my fork. Hopefully after merging it will be enabled also in this repository.

I totally agree.

**Review by traversaro — APPROVED** [review]

**Comment by diegoferigo** [comment]
Travis failures are not related with this PR (somehow `ccache` is not found in the macOS PATH despite ccache support is enabled).

The culprit [seems to be](https://travis-ci.org/robotology/how-to-export-cpp-library/jobs/575544270#L111) a missing `brew update` execution. This goes beyond the scope of this PR though.

**Review by claudiofantacci — APPROVED** [review]
This is great! Thanks 🚀

**Comment by diegoferigo** [comment]
Thanks for merging! Everything seems working as expected, badge included :heart_eyes:

**Pull request merged**